### PR TITLE
MINOR: Better messaging for invalid fetch response

### PR DESCRIPTION
**Pull request by jsancio**
Users have reported (KAFKA-7565) that when consumer poll wake up is used,
it is possible to receive fetch responses that don't match the copied topic
partitions collection for the session when the fetch request was created.

This commit improves the error handling here by throwing an
IllegalStateException instead of a NullPointerException. And by
generating a message for the exception that includes a bit of more
information.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Comment by jsancio** [comment]
cc @hachikuji @cmccabe

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java [hunk 1]
```diff
@@ -241,13 +241,19 @@ public void onSuccess(ClientResponse resp) {
...
-                                    FetchResponse.PartitionData<Records> fetchData = entry.getValue();
-
-                                    log.debug("Fetch {} at offset {} for partition {} returned fetch data {}",
-                                            isolationLevel, fetchOffset, partition, fetchData);
-                                    completedFetches.add(new CompletedFetch(partition, fetchOffset, fetchData, metricAggregator,
-                                            resp.requestHeader().apiVersion()));
+                                    FetchRequest.PartitionData requestData = data.sessionPartitions().get(partition);
+                                    if (requestData == null) {
+                                        // Received fetch response for missing session partition
+                                        throw new IllegalStateException("Received response for partition " + partition + " which is not part of the session");
```
Comment (hachikuji) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java]: I wonder if we could include any additional state from the session itself to get closer to the root of the problem. For example, was this a response to a full fetch request?

Comment (jsancio) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java]: Add more information. Read the new code twice as none of the tests exercise the formatting logic.

**Review by hachikuji — APPROVED** [review]
LGTM. I'd suggest changing this to a MINOR PR since we are not addressing the root cause of KAFKA-7565.

**Comment by jsancio** [comment]
> LGTM. I'd suggest changing this to a MINOR PR since we are not addressing the root cause of KAFKA-7565.

Done.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java [hunk 2]
```diff
@@ -241,13 +242,30 @@ public void onSuccess(ClientResponse resp) {
...
-                                            isolationLevel, fetchOffset, partition, fetchData);
-                                    completedFetches.add(new CompletedFetch(partition, fetchOffset, fetchData, metricAggregator,
-                                            resp.requestHeader().apiVersion()));
+                                    FetchRequest.PartitionData requestData = data.sessionPartitions().get(partition);
+                                    if (requestData == null) {
+                                        String message;
+                                        if (data.metadata().isFull()) {
+                                            message = MessageFormatter.arrayFormat(
+                                                    "Response for missing full request partition: partition={}; metadata={}",
+                                                    new Object[]{partition, data.metadata()}).getMessage();
```
Comment (ijuma) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java:251]: Any reason why we didn't use String.format here?

Comment (jsancio) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java:251]: No technical reason. I see that `String.format` is used in a few places already. Let me know if you would like me to change this.

Comment (jsancio) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java:251]: @ijuma, Actually, I now remember why I did it. `MessageFormatter` supports Java arrays while `String.format` does not.

Comment (ijuma) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java:251]: Why would you want arrays instead of varargs though?

Comment (jsancio) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java:251]:
```
String.format("array=%s", new Object[]{1,2,3})
array=[I@76f78e3a

MessageFormatter.arrayFormat("array={}", new Object[]{new Object[]{1,2,3}}).getMessage()
array=[1,2,3]
```
Right?

**Pull request merged**